File: python/src/byoc/providers/gdrive/adapter.py
"""Google Drive provider adapter."""

from __future__ import annotations

from collections.abc import AsyncIterator, Awaitable, Callable
from datetime import datetime, timedelta, timezone
from typing import TypeVar

import httpx

from ...errors import InvalidInputError, ObjectNotFoundError
from ...paths import get_basename, normalize_virtual_path
from ...types import (
    ProviderCapabilities,
    ProviderManifest,
    StorageInput,
    StorageObject,
    StorageOutput,
    StorageQuota,
    UploadGrant,
    UploadGrantOptions,
    UploadOptions,
)
from ._http import RESUMABLE_THRESHOLD, DriveHttpClient
from ._oauth import GoogleDriveScope, GoogleOAuthClient
from ._resolver import DrivePathResolver, LruTtlPathCache
from ._tokens import TokenSession, TokenStorage

PROVIDER_ID = "google-drive"

T = TypeVar("T")


class GoogleDriveProvider:
    """Store application files in the Drive account your user already owns.

    Uses the ``drive.file`` scope by default, so the app can only see files it
    created or that the user explicitly opened with it -- which keeps you out of
    Google's Restricted Scope security assessment.

    Args:
        client_id: OAuth client id.
        client_secret: Only for confidential server-side clients.
        redirect_uri: Redirect registered with the OAuth client.
        scopes: Defaults to ``[GoogleDriveScope.FILE]``.
        token_storage: Where the session persists between runs.
        session: An existing session, for pre-authenticated callers.
        root_folder_name: App folder created at the Drive root.
        cache: Optional path-cache override.
        client: An ``httpx.AsyncClient`` to reuse.
    """

    def __init__(
        self,
        *,
        client_id: str,
        client_secret: str | None = None,
        redirect_uri: str | None = None,
        scopes: list[str] | None = None,
        token_storage: TokenStorage | None = None,
        session: TokenSession | None = None,
        root_folder_name: str = "BYOC",
        cache: LruTtlPathCache | None = None,
        client: httpx.AsyncClient | None = None,
    ) -> None:
        if not client_id:
            raise InvalidInputError(
                "GoogleDriveProvider requires a 'client_id'.", provider=PROVIDER_ID
            )

        self.oauth = GoogleOAuthClient(
            client_id=client_id,
            client_secret=client_secret,
            redirect_uri=redirect_uri,
            scopes=scopes or [GoogleDriveScope.FILE],
            token_storage=token_storage,
            session=session,
            client=client,
        )
        self.http = DriveHttpClient(self.oauth.get_access_token, client=client)
        self.resolver = DrivePathResolver(
            self.http, root_folder_name=root_folder_name, cache=cache
        )
        self.root_folder_name = root_folder_name

    # -- lifecycle ---------------------------------------------------------

    def manifest(self) -> ProviderManifest:
        return ProviderManifest(
            id=PROVIDER_ID,
            name="Google Drive",
            category="personal-cloud",
            authentication="oauth2",
            supports_user_owned_storage=True,
            adapter_version="0.4.0",
        )

    def capabilities(self) -> ProviderCapabilities:
        return ProviderCapabilities(
            folders=True,
            sharing=True,
            public_urls=False,
            resumable_uploads=True,
            versioning=True,
            quota=True,
            server_side_copy=True,
            direct_upload=True,
        )

    async def connect(self) -> None:
        """Verify the session and ensure the app root folder exists."""
        await self.resolver.ensure_root_folder()

    async def disconnect(self) -> None:
        """Close HTTP resources. Does not revoke the token -- use ``oauth.revoke()``."""
        await self.http.aclose()
        await self.oauth.aclose()

    # -- operations --------------------------------------------------------

    async def upload(
        self, path: str, data: StorageInput, options: UploadOptions | None = None
    ) -> StorageObject:
        normalized = normalize_virtual_path(path)
        if not normalized:
            raise InvalidInputError("Upload requires a file path.", provider=PROVIDER_ID)

        payload: bytes | None
        if isinstance(data, str):
            payload = data.encode("utf-8")
        elif isinstance(data, bytes | bytearray | memoryview):
            payload = bytes(data)
        else:
            # An async iterator streams into the resumable session, so the
            # object never has to fit in memory.
            payload = None
            stream = data

        mime = (options.mime_type if options else None) or "application/octet-stream"
        parent_id = await self.resolver.resolve_parent_folder_id(normalized, create=True)

        metadata = self.http.build_metadata(
            name=get_basename(normalized),
            parent_id=parent_id,
            virtual_path=normalized,
            mime_type=mime,
            extra=dict(options.metadata) if options else None,
        )

        resumable = options.resumable if options and options.resumable is not None else None
        wants_resumable = (
            # A stream has no length to compare, and chunked transfer is the
            # only way to send one, so it is always resumable.
            True
            if payload is None
            else resumable
            if resumable is not None
            else (len(payload) > RESUMABLE_THRESHOLD or bool(options and options.on_progress))
        )

        chunk_size = options.chunk_size if options and options.chunk_size else 8 * 1024 * 1024

        if payload is None:
            upload_url = await self.http.start_resumable_upload(metadata, mime)
            resource = await self.http.stream_chunks(
                upload_url,
                stream,
                chunk_size=chunk_size,
                on_progress=options.on_progress if options else None,
            )
        elif wants_resumable:
            upload_url = await self.http.start_resumable_upload(metadata, mime)
            resource = await self.http.upload_chunks(
                upload_url,
                payload,
                chunk_size=chunk_size,
                on_progress=options.on_progress if options else None,
            )
        else:
            resource = await self.http.multipart_upload(metadata, payload, mime)

        obj = self.http.to_storage_object(resource, normalized)
        self.resolver.cache.set(normalized, obj.provider_id)
        return obj

    async def download(self, path: str) -> StorageOutput:
        normalized = normalize_virtual_path(path)

        async def fetch(file_id: str) -> tuple[dict[str, object], bytes]:
            return await self.http.get_file(file_id), await self.http.download_file(file_id)

        resource, body = await self._with_healing(normalized, fetch)

        async def stream() -> AsyncIterator[bytes]:
            yield body

        async def read() -> bytes:
            return body

        return StorageOutput(
            metadata=self.http.to_storage_object(resource, normalized), stream=stream, read=read
        )

    async def delete(self, path: str) -> None:
        normalized = normalize_virtual_path(path)
        try:
            await self._with_healing(normalized, self.http.delete_file)
        except ObjectNotFoundError:
            return  # Already gone: deletion is idempotent.
        finally:
            self.resolver.invalidate(normalized)

    async def metadata(self, path: str) -> StorageObject:
        normalized = normalize_virtual_path(path)
        resource = await self._with_healing(normalized, self.http.get_file)
        return self.http.to_storage_object(resource, normalized)

    async def exists(self, path: str) -> bool:
        try:
            await self.metadata(path)
        except ObjectNotFoundError:
            return False
        return True

    async def list(self, path: str | None = None) -> list[StorageObject]:
        normalized = normalize_virtual_path(path)
        parent_id = await self.resolver.resolve_folder_id(normalized, create=False)

        query = f"'{parent_id}' in parents and trashed = false"
        resources = await self.http.list_files(query, page_size=1000)

        results: list[StorageObject] = []
        for resource in resources:
            name = str(resource.get("name", ""))
            child_path = f"{normalized}/{name}" if normalized else name
            results.append(self.http.to_storage_object(resource, child_path))
        return results

    async def create_folder(self, path: str) -> StorageObject:
        normalized = normalize_virtual_path(path)
        if not normalized:
            raise InvalidInputError("Create folder requires a path.", provider=PROVIDER_ID)

        folder_id = await self.resolver.resolve_folder_id(normalized, create=True)
        return StorageObject(
            id=f"gdrive_{folder_id}",
            path=normalized,
            name=get_basename(normalized),
            provider=PROVIDER_ID,
            provider_id=folder_id,
            type="folder",
        )

    async def create_upload_grant(
        self, path: str, options: UploadGrantOptions | None = None
    ) -> UploadGrant:
        """Open a resumable session and hand back its URI as an upload grant.

        Unlike S3's signed URL this is not a signature over a path -- Drive has
        already created the pending file, and the session URI is the only way
        to write to it. Two consequences worth knowing:

        - ``size_bytes`` is required. Drive wants ``X-Upload-Content-Length``
          up front, and a session opened with the wrong total rejects the
          final chunk.
        - Sessions last about a week, far longer than a signed URL, so this
          still honours ``expires_in_seconds`` and reports the earlier of the two.
        """
        resolved = options or UploadGrantOptions()
        normalized = normalize_virtual_path(path)

        if resolved.size_bytes is None:
            raise InvalidInputError(
                "Google Drive needs the total size before opening a resumable session. "
                "Pass size_bytes (file.size in the browser) when creating the grant.",
                provider=PROVIDER_ID,
            )

        parent_id = await self.resolver.resolve_parent_folder_id(normalized, create=True)
        mime_type = resolved.mime_type or "application/octet-stream"

        session_uri = await self.http.start_resumable_upload(
            {
                "name": get_basename(normalized),
                "parents": [parent_id],
                "mimeType": mime_type,
                "appProperties": {"byocVirtualPath": normalized},
            },
            mime_type,
        )

        return UploadGrant(
            provider=PROVIDER_ID,
            path=normalized,
            url=session_uri,
            method="PUT",
            headers={},
            protocol="resumable",
            expires_at=datetime.now(timezone.utc)
            + timedelta(seconds=resolved.expires_in_seconds),
            # Drive requires every chunk but the last to be a multiple of 256 KiB.
            chunk_size=8 * 1024 * 1024,
            max_bytes=resolved.size_bytes,
        )

    async def copy(self, source: str, destination: str) -> None:
        """Server-side copy into ``destination``, creating parents as needed."""
        source_norm = normalize_virtual_path(source)
        dest_norm = normalize_virtual_path(destination)
        if not source_norm or not dest_norm:
            raise InvalidInputError(
                "Copy requires both a source and a destination path.", provider=PROVIDER_ID
            )

        source_id = await self.resolver.resolve_file_id(source_norm)
        dest_parent_id = await self.resolver.resolve_parent_folder_id(dest_norm, create=True)

        copied = await self.http.copy_file(source_id, get_basename(dest_norm), dest_parent_id)
        self.resolver.cache.set(dest_norm, str(copied["id"]))

    async def move(self, source: str, destination: str) -> None:
        """Reparent the file server-side, so no bytes are transferred."""
        source_norm = normalize_virtual_path(source)
        dest_norm = normalize_virtual_path(destination)
        if not source_norm or not dest_norm:
            raise InvalidInputError(
                "Move requires both a source and a destination path.", provider=PROVIDER_ID
            )

        source_id = await self.resolver.resolve_file_id(source_norm)
        old_parent_id = await self.resolver.resolve_parent_folder_id(source_norm, create=False)
        new_parent_id = await self.resolver.resolve_parent_folder_id(dest_norm, create=True)

        await self.http.move_file(
            source_id, new_parent_id, old_parent_id, get_basename(dest_norm)
        )
        # The old path no longer resolves; the new one now points at the same id.
        self.resolver.invalidate(source_norm)
        self.resolver.cache.set(dest_norm, source_id)

    async def quota(self) -> StorageQuota:
        return await self.http.get_quota()

    # -- internals ---------------------------------------------------------

    async def _with_healing(
        self, normalized: str, operation: Callable[[str], Awaitable[T]]
    ) -> T:
        """Resolve a path and run ``operation`` on it, healing a stale cache.

        Users rename and delete files in the Drive web UI, so a cached id can go
        bad in two different ways: the path stops resolving, or it resolves to
        an id Drive has already deleted. Only the first is caught by retrying
        resolution alone, which is why the operation itself is inside the retry.
        A second failure is genuine and propagates.
        """
        try:
            return await operation(await self.resolver.resolve_file_id(normalized))
        except ObjectNotFoundError:
            self.resolver.invalidate(normalized)
            return await operation(await self.resolver.resolve_file_id(normalized))

File: packages/google-drive/src/adapter.ts
import type {
  BYOCProvider,
  ProviderCapabilities,
  ProviderManifest,
  StorageInput,
  StorageObject,
  StorageOutput,
  StorageQuota,
  UploadOptions,
  BackupOptions,
  UploadGrant,
  UploadGrantOptions
} from "@byoc/core";
import {
  BYOCErrorCode,
  StorageError,
  normalizeVirtualPath,
  getBasename
} from "@byoc/core";
import type { GoogleDriveProviderConfig } from "./auth/types.js";
import { GoogleOAuthClient } from "./auth/oauth-client.js";
import { DriveHttpClient } from "./api/http.js";
import {
  GoogleDrivePathResolver,
  LruTtlPathCache,
  type PathCache
} from "./paths/virtual-path.js";

import { ResumableUploader } from "./api/resumable.js";
import { buildMultipartBody } from "./api/multipart.js";

/**
 * GoogleDriveProvider — Official reference adapter for Google Drive.
 */
export class GoogleDriveProvider implements BYOCProvider {
  public readonly config: GoogleDriveProviderConfig;
  public readonly oauth: GoogleOAuthClient;
  public readonly http: DriveHttpClient;
  public readonly uploader: ResumableUploader;
  public readonly resolver: GoogleDrivePathResolver;

  constructor(
    config: GoogleDriveProviderConfig,
    options?: { pathCache?: PathCache }
  ) {
    if (!config || !config.auth || !config.auth.clientId) {
      throw new StorageError({
        code: BYOCErrorCode.INVALID_INPUT,
        message: "GoogleDriveProvider requires a valid 'auth.clientId'.",
        provider: "google-drive",
        retryable: false
      });
    }

    this.config = {
      ...config,
      rootFolderName: config.rootFolderName ?? "BYOC"
    };

    this.oauth = new GoogleOAuthClient(config.auth);
    this.http = new DriveHttpClient(this.oauth);
    this.uploader = new ResumableUploader(this.oauth);
    const cache = options?.pathCache ?? new LruTtlPathCache();
    this.resolver = new GoogleDrivePathResolver(this.http, this.config.rootFolderName, cache);
  }

  /**
   * Manifest identifying this provider to the BYOC runtime.
   */
  public manifest(): ProviderManifest {
    return {
      id: "google-drive",
      name: "Google Drive",
      category: "personal-cloud",
      authentication: "oauth2",
      supportsUserOwnedStorage: true,
      adapterVersion: "0.3.0"
    };
  }

  /**
   * Declares the capabilities supported by Google Drive.
   */
  public capabilities(): ProviderCapabilities {
    return {
      folders: true,
      sharing: true,
      publicUrls: false,
      resumableUploads: true,
      versioning: false,
      quota: true,
      serverSideCopy: true,
      directUpload: true
    };
  }

  /**
   * Initialize and verify OAuth connection session, ensuring the root app folder exists.
   */
  public async connect(): Promise<void> {
    const hasSession = await this.oauth.hasValidSession();
    if (!hasSession) {
      throw new StorageError({
        code: BYOCErrorCode.AUTH_REQUIRED,
        message: "Google Drive authentication required: No valid session found. Please complete OAuth flow.",
        provider: "google-drive",
        statusCode: 401,
        retryable: false
      });
    }

    // Verify token validity by resolving root folder
    await this.resolver.ensureRootFolder();
  }

  /**
   * Disconnect and clear local token/cache sessions, revoking active credentials.
   */
  public async disconnect(): Promise<void> {
    await this.oauth.revoke();
    await this.resolver.clearCache();
  }

  /**
   * Uploads a file to Google Drive (using multipart for small files, resumable for large files).
   */
  public async upload(
    path: string,
    data: StorageInput,
    options?: UploadOptions
  ): Promise<StorageObject> {
    const normalized = normalizeVirtualPath(path);
    const filename = getBasename(normalized);
    const parentFolderId = await this.resolver.resolveParentFolderId(normalized);

    const metadata = {
      name: filename,
      parents: [parentFolderId],
      mimeType: options?.mimeType,
      appProperties: {
        byocVirtualPath: normalized,
        ...((options?.metadata as Record<string, string>) || {})
      }
    };

    // Determine if resumable upload is appropriate (explicit flag, onProgress listener, or large byte length)
    const byteLength =
      typeof data === "string"
        ? Buffer.byteLength(data)
        : data instanceof Uint8Array || (typeof Buffer !== "undefined" && Buffer.isBuffer(data))
          ? data.byteLength
          : undefined;

    const isLarge = byteLength !== undefined && byteLength >= 5 * 1024 * 1024; // 5 MB threshold
    const shouldUseResumable = options?.resumable || options?.onProgress || isLarge;

    // A stream has no length to compare against a threshold, and chunked
    // transfer is the only way to send one, so it always goes resumable.
    const isStream =
      byteLength === undefined &&
      typeof data !== "string" &&
      typeof (data as AsyncIterable<Uint8Array>)?.[Symbol.asyncIterator] === "function";

    let resource;
    if (isStream) {
      resource = await this.uploader.uploadStream(
        metadata,
        data as AsyncIterable<Uint8Array>,
        options?.mimeType ?? "application/octet-stream",
        options?.chunkSize,
        options?.onProgress
      );
    } else if (shouldUseResumable) {
      resource = await this.uploader.upload(metadata, data, {
        mimeType: options?.mimeType,
        chunkSize: options?.chunkSize,
        onProgress: options?.onProgress
      });
    } else {
      const multipart = await buildMultipartBody(metadata, data, options?.mimeType);
      const boundaryMatch = multipart.contentType.match(/boundary=(.+)$/);
      const boundary = boundaryMatch ? boundaryMatch[1]! : `BYOC_${Date.now()}`;
      resource = await this.http.uploadMultipart(boundary, multipart.body);
    }

    // Cache the resolved file ID
    await this.resolver.setCached(normalized, resource.id);

    return this.http.toStorageObject(resource, normalized);
  }

  /**
   * Downloads a file from Google Drive with 404 cache self-healing.
   */
  public async download(path: string): Promise<StorageOutput> {
    const normalized = normalizeVirtualPath(path);

    try {
      const fileId = await this.resolver.resolveFileId(normalized);
      const resource = await this.http.getFile(fileId);
      const media = await this.http.downloadMedia(fileId);

      return {
        stream: media.stream,
        arrayBuffer: media.arrayBuffer,
        text: media.text,
        metadata: this.http.toStorageObject(resource, normalized)
      };
    } catch (err) {
      // If 404 occurs on cached ID, invalidate cache and retry once
      if (StorageError.isStorageError(err) && err.code === BYOCErrorCode.OBJECT_NOT_FOUND) {
        await this.resolver.invalidate(normalized);
        const freshFileId = await this.resolver.resolveFileId(normalized, { skipCache: true });
        const resource = await this.http.getFile(freshFileId);
        const media = await this.http.downloadMedia(freshFileId);

        return {
          stream: media.stream,
          arrayBuffer: media.arrayBuffer,
          text: media.text,
          metadata: this.http.toStorageObject(resource, normalized)
        };
      }
      throw err;
    }
  }

  /**
   * Deletes an object from Google Drive with cache invalidation.
   * If softDelete option is specified, moves the file to Google Drive Trash.
   */
  public async delete(path: string, options?: { softDelete?: boolean }): Promise<void> {
    const normalized = normalizeVirtualPath(path);
    const fileId = await this.resolver.resolveFileId(normalized);
    if (options?.softDelete) {
      await this.http.trashFile(fileId);
    } else {
      await this.http.deleteFile(fileId);
    }
    await this.resolver.invalidate(normalized);
  }

  /**
   * Server-side instant copy in Google Drive using files.copy without re-uploading bytes.
   */
  /**
   * Opens a resumable session and hands back its URI as an upload grant.
   *
   * Unlike S3's signed URL, this is not a signature over a path -- Drive has
   * already created the pending file, and the session URI is the only way to
   * write to it. Two consequences worth knowing:
   *
   *  - `sizeBytes` is required. Drive wants X-Upload-Content-Length up front,
   *    and a session opened with the wrong total rejects the final chunk.
   *  - Sessions last about a week, far longer than a signed URL, so this
   *    still honours `expiresInSeconds` and reports the earlier of the two.
   */
  public async createUploadGrant(
    path: string,
    options: UploadGrantOptions = {}
  ): Promise<UploadGrant> {
    const normalized = normalizeVirtualPath(path);
    if (options.sizeBytes === undefined) {
      throw new StorageError({
        code: BYOCErrorCode.INVALID_INPUT,
        message:
          "Google Drive needs the total size before opening a resumable session. " +
          "Pass sizeBytes (file.size in the browser) when creating the grant.",
        provider: "google-drive",
        retryable: false
      });
    }

    const parentFolderId = await this.resolver.resolveParentFolderId(normalized);
    const mimeType = options.mimeType ?? "application/octet-stream";

    const sessionUri = await this.uploader.createSession(
      {
        name: getBasename(normalized),
        parents: [parentFolderId],
        mimeType,
        appProperties: { byocVirtualPath: normalized }
      },
      options.sizeBytes,
      mimeType
    );

    const expiresInSeconds = options.expiresInSeconds ?? 900;
    return {
      provider: "google-drive",
      path: normalized,
      url: sessionUri,
      method: "PUT",
      headers: {},
      protocol: "resumable",
      expiresAt: new Date(Date.now() + expiresInSeconds * 1000),
      // Drive requires every chunk but the last to be a multiple of 256 KiB.
      chunkSize: 8 * 1024 * 1024,
      maxBytes: options.sizeBytes
    };
  }

  public async copy(source: string, destination: string): Promise<void> {
    const sourceNorm = normalizeVirtualPath(source);
    const destNorm = normalizeVirtualPath(destination);

    const sourceFileId = await this.resolver.resolveFileId(sourceNorm);
    const destParentId = await this.resolver.resolveParentFolderId(destNorm);
    const newFilename = getBasename(destNorm);

    const copied = await this.http.copyFile(sourceFileId, newFilename, destParentId);
    await this.resolver.setCached(destNorm, copied.id);
  }

  /**
   * Server-side instant move in Google Drive using files.update (parent swapping) without re-uploading bytes.
   */
  public async move(source: string, destination: string): Promise<void> {
    const sourceNorm = normalizeVirtualPath(source);
    const destNorm = normalizeVirtualPath(destination);

    const sourceFileId = await this.resolver.resolveFileId(sourceNorm);
    const oldParentId = await this.resolver.resolveParentFolderId(sourceNorm);
    const newParentId = await this.resolver.resolveParentFolderId(destNorm);
    const newFilename = getBasename(destNorm);

    await this.http.moveFile(sourceFileId, newParentId, oldParentId, newFilename);
    await this.resolver.invalidate(sourceNorm);
    await this.resolver.setCached(destNorm, sourceFileId);
  }

  /**
   * Lists files and folders within a path.
   */
  public async list(path?: string): Promise<StorageObject[]> {
    const normalized = normalizeVirtualPath(path);
    const parentId = normalized
      ? await this.resolver.resolveFileId(normalized)
      : await this.resolver.ensureRootFolder();

    const query = `'${parentId}' in parents and trashed = false`;
    const response = await this.http.listFiles(query);

    return response.files.map((file) => {
      const childPath = normalized ? `${normalized}/${file.name}` : file.name;
      return this.http.toStorageObject(file, childPath);
    });
  }

  /**
   * Checks if an object exists in Google Drive.
   */
  public async exists(path: string): Promise<boolean> {
    const normalized = normalizeVirtualPath(path);
    if (!normalized) return true;

    try {
      await this.resolver.resolveFileId(normalized);
      return true;
    } catch (err) {
      if (StorageError.isStorageError(err) && err.code === BYOCErrorCode.OBJECT_NOT_FOUND) {
        return false;
      }
      throw err;
    }
  }

  /**
   * Retrieves metadata for an object in Google Drive.
   */
  public async metadata(path: string): Promise<StorageObject> {
    const normalized = normalizeVirtualPath(path);
    const fileId = await this.resolver.resolveFileId(normalized);
    const resource = await this.http.getFile(fileId);
    return this.http.toStorageObject(resource, normalized);
  }

  /**
   * Creates a virtual folder in Google Drive.
   */
  public async createFolder(path: string): Promise<StorageObject> {
    const normalized = normalizeVirtualPath(path);
    const folderId = await this.resolver.resolveOrCreateFolderPath(normalized);
    const resource = await this.http.getFile(folderId);
    return this.http.toStorageObject(resource, normalized);
  }

  /**
   * Retrieves user storage quota from Google Drive.
   */
  public async quota(): Promise<StorageQuota> {
    const about = await this.http.getAboutQuota();
    const quotaData = about.storageQuota;

    return {
      total: quotaData?.limit ? Number(quotaData.limit) : undefined,
      used: quotaData?.usage ? Number(quotaData.usage) : 0,
      available:
        quotaData?.limit && quotaData.usage
          ? Math.max(0, Number(quotaData.limit) - Number(quotaData.usage))
          : undefined
    };
  }

  /**
   * Specialized backup helper.
   */
  public async backup(
    payload: StorageInput,
    options?: BackupOptions
  ): Promise<StorageObject> {
    const folder = normalizeVirtualPath(options?.folder ?? "Backups");
    const filename =
      options?.filename ?? `backup-${new Date().toISOString().replace(/[:.]/g, "-")}.json`;
    const targetPath = folder ? `${folder}/${filename}` : filename;

    return this.upload(targetPath, payload, {
      mimeType: options?.mimeType ?? "application/json",
      onProgress: options?.onProgress
    });
  }
}
